File: frontend/src/types/job.ts
export type AssetKind = "clip" | "thumbnail";

export interface UploadAsset {
    kind: AssetKind;
    filename: string;
    content_type?: string | null;
    size_bytes?: number | null;
    sequence?: number | null;
}

export interface CreateJobRequest {
    title: string;
    description: string;
    assets: UploadAsset[];
}

export interface CreateJobResponse {
    job_id: string;
    status: string;
}

export interface AddAssetsRequest {
    assets: UploadAsset[];
}

export interface UploadInstruction {
    filename: string;
    kind: AssetKind;
    url: string;
    method: string;
    headers: Record<string, string>;
}

export interface AddAssetsResponse {
    job_id: string;
    uploads: UploadInstruction[];
}

export interface CompleteJobResponse {
    job_id: string;
    status: string;
    queue_message: {
        job_id: string;
        title: string;
        description: string;
        clips: {
            filename: string;
            sequence?: number | null;
        }[];
        thumbnail: {
            filename: string;
        } | null;
    };
}

export interface JobStatusResponse {
    job_id: string;
    status: string;
    title: string;
    description: string;
    assets: UploadAsset[];
    youtube_url?: string | null;
    error?: string | null;
}

File: backend/app/schemas/job.py
from __future__ import annotations

from pydantic import BaseModel, Field
from typing import List, Literal, Optional


AssetKind = Literal["clip", "thumbnail"]


class UploadAsset(BaseModel):
    kind: AssetKind = "clip"
    filename: str
    content_type: Optional[str] = None 
    size_bytes: Optional[int] = None
    sequence: Optional[int] = None


class CreateJobRequest(BaseModel):
    title: str
    description: str = ""
    assets: List[UploadAsset] = Field(default_factory=list)


class CreateJobResponse(BaseModel):
    job_id: str
    status: str


class AddAssetsRequest(BaseModel):
    assets: List[UploadAsset]

class UpdateJobStatusRequest(BaseModel):
    status: str
    youtube_video_id: Optional[str] = None

class UploadInstruction(BaseModel):
    filename: str
    kind: AssetKind
    url: str 
    method: str = "PUT"
    headers: dict = Field(default_factory=dict)


class AddAssetsResponse(BaseModel):
    job_id: str
    uploads: List[UploadInstruction]


class SubmitJobResponse(BaseModel):
    job_id: str
    status: str
    queue_message: dict 

class JobProgress(BaseModel):
    rendering_progress: int = 0
    publishing_progress: int = 0

class JobStatusResponse(BaseModel):
    job_id: str
    created_at: Optional[str] = None
    status: str
    title: str
    description: str
    assets: List[UploadAsset]
    youtube_video_id: Optional[str] = None
    error: Optional[str] = None
    progress: Optional[JobProgress] = None


class UpdateJobProgressRequest(BaseModel):
    rendering_progress: Optional[int] = None
    publishing_progress : Optional[int] = None